File: typescript/src/policy.ts
import type { AgentProfile, Adjustment, JsonObject, PermissionDecision } from "./types.js";

const DECISION_RANK: Readonly<Record<PermissionDecision, number>> = { deny: 0, ask: 1, allow: 2 };

export function narrowDecision(policy: PermissionDecision, requested: PermissionDecision): PermissionDecision {
  return DECISION_RANK[policy] <= DECISION_RANK[requested] ? policy : requested;
}

function wildcard(pattern: string): RegExp {
  const escaped = pattern.replace(/[.+?^${}()|[\]\\]/g, "\\$&").replaceAll("**", "\u0000").replaceAll("*", "[^/]*").replaceAll("\u0000", ".*");
  return new RegExp(`^${escaped}$`);
}

function matchesAny(value: string, patterns: string[]): boolean {
  return patterns.some((pattern) => wildcard(pattern).test(value));
}

export interface EffectiveTools {
  tools: string[];
  adjustments: Adjustment[];
}

export function intersectTools(profile: AgentProfile | JsonObject, grantedTools: Iterable<string>): EffectiveTools {
  const spec = (profile.spec as JsonObject | undefined) ?? {};
  const tools = (spec.tools as JsonObject | undefined) ?? {};
  const policy = typeof tools.policy === "string" ? tools.policy : "allowlist";
  const allow = Array.isArray(tools.allow) ? tools.allow.filter((value): value is string => typeof value === "string") : [];
  const deny = Array.isArray(tools.deny) ? tools.deny.filter((value): value is string => typeof value === "string") : [];
  const granted = [...grantedTools];
  const requested = policy === "inherit" ? granted : policy === "denylist" ? granted : granted.filter((tool) => matchesAny(tool, allow));
  const effective = requested.filter((tool) => !matchesAny(tool, deny));
  const dropped = granted.filter((tool) => !effective.includes(tool));
  const adjustments: Adjustment[] = dropped.map((tool) => ({
    field: `spec.tools:${tool}`,
    requested: tool,
    effective: null,
    reason: deny.includes(tool) || matchesAny(tool, deny) ? "profile deny rule" : "not included by profile allowlist",
  }));
  return { tools: effective.sort(), adjustments };
}

export function narrowPermissionMap(
  requested: Record<string, PermissionDecision>,
  policy: Record<string, PermissionDecision>,
): { permissions: Record<string, PermissionDecision>; adjustments: Adjustment[] } {
  const permissions: Record<string, PermissionDecision> = {};
  const adjustments: Adjustment[] = [];
  for (const key of new Set([...Object.keys(requested), ...Object.keys(policy)])) {
    const wanted = requested[key] ?? "ask";
    const ceiling = policy[key] ?? "ask";
    const effective = narrowDecision(ceiling, wanted);
    permissions[key] = effective;
    if (effective !== wanted) adjustments.push({ field: `spec.permissions.${key}`, requested: wanted, effective, reason: "local policy ceiling" });
  }
  return { permissions, adjustments };
}

File: rust/src/policy.rs
use std::collections::BTreeSet;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{Adjustment, AgentProfile, object, strings};

/// Ordered permission decision used for fail-closed narrowing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PermissionDecision {
    /// Operation is forbidden.
    Deny,
    /// Operation requires approval.
    Ask,
    /// Operation is allowed.
    Allow,
}

/// Returns the more restrictive of a policy ceiling and requested decision.
pub fn narrow_decision(
    policy: PermissionDecision,
    requested: PermissionDecision,
) -> PermissionDecision {
    std::cmp::min(policy, requested)
}

/// Effective tool set and explanations for removed grants.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EffectiveTools {
    /// Sorted, deduplicated tools allowed by both profile and harness.
    pub tools: Vec<String>,
    /// Explanations for requested tools that were removed.
    pub adjustments: Vec<Adjustment>,
}

fn wildcard(pattern: &str, value: &str) -> bool {
    if pattern == "*" {
        return true;
    }
    if let Some(prefix) = pattern.strip_suffix('*') {
        return value.starts_with(prefix);
    }
    pattern == value
}

/// Intersects harness-granted tools with an OAP profile's tool policy.
pub fn intersect_tools(
    profile: &AgentProfile,
    granted: impl IntoIterator<Item = impl AsRef<str>>,
) -> EffectiveTools {
    let tools = object(object(profile.get("spec")).get("tools"));
    let policy = tools
        .get("policy")
        .and_then(Value::as_str)
        .unwrap_or("inherit");
    let allow = strings(tools.get("allow"));
    let deny = strings(tools.get("deny"));
    let mut effective = vec![];
    let mut adjustments = vec![];
    for raw in granted {
        let tool = raw.as_ref();
        let requested = policy != "deny_all"
            && (policy != "allowlist" || allow.iter().any(|item| wildcard(item, tool)))
            && !deny.iter().any(|item| wildcard(item, tool));
        if requested {
            effective.push(tool.to_owned());
        } else {
            adjustments.push(Adjustment {
                field: format!("tools.{tool}"),
                requested: Value::String("allow".into()),
                effective: Value::String("deny".into()),
                reason: "profile and harness capabilities intersect; they never union".into(),
            });
        }
    }
    effective.sort();
    effective.dedup();
    EffectiveTools {
        tools: effective,
        adjustments,
    }
}

/// Applies field-by-field harness ceilings to a requested permission map.
pub fn narrow_permission_map(
    requested: &std::collections::BTreeMap<String, PermissionDecision>,
    policy: &std::collections::BTreeMap<String, PermissionDecision>,
) -> (
    std::collections::BTreeMap<String, PermissionDecision>,
    Vec<Adjustment>,
) {
    let keys: BTreeSet<_> = requested.keys().chain(policy.keys()).cloned().collect();
    let mut effective = std::collections::BTreeMap::new();
    let mut adjustments = vec![];
    for key in keys {
        let ask = *requested.get(&key).unwrap_or(&PermissionDecision::Ask);
        let ceiling = *policy.get(&key).unwrap_or(&PermissionDecision::Ask);
        let value = narrow_decision(ceiling, ask);
        effective.insert(key.clone(), value);
        if value != ask {
            adjustments.push(Adjustment {
                field: key,
                requested: serde_json::to_value(ask).unwrap(),
                effective: serde_json::to_value(value).unwrap(),
                reason: "harness policy is the upper bound".into(),
            });
        }
    }
    (effective, adjustments)
}
